cirq-core/cirq/ops/conditional_operation_test.py

```diff
@@ -0,0 +1,353 @@
+    circuit = cirq.Circuit(
+        cirq.measure(q0, key='a'),
+        cirq.measure(q0, key='b'),
+        cirq.X(q1).with_conditions('a', 'b'),
+    )
+
+    cirq.testing.assert_has_diagram(
+        circuit,
+        """
+0: ───M───M───────
+      ║   ║
+1: ───╫───╫───X───
+      ║   ║   ║
+a: ═══@═══╬═══^═══
+          ║   ║
+b: ═══════@═══^═══
+""",
+        use_unicode_characters=True,
+    )
+
+
+def test_diagram_multiple_ops_single_moment():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(
+        cirq.measure(q0, key='a'),
+        cirq.measure(q1, key='b'),
+        cirq.X(q0).with_conditions('a'),
+        cirq.X(q1).with_conditions('b'),
+    )
+
+    cirq.testing.assert_has_diagram(
+        circuit,
+        """
+      ┌──┐   ┌──┐
+0: ────M──────X─────
+       ║      ║
+1: ────╫M─────╫X────
+       ║║     ║║
+a: ════@╬═════^╬════
+        ║      ║
+b: ═════@══════^════
+      └──┘   └──┘
+""",
+        use_unicode_characters=True,
+    )
+
+
+def test_diagram_subcircuit():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(
+        cirq.CircuitOperation(
+            cirq.FrozenCircuit(
+                cirq.measure(q0, key='a'),
+                cirq.X(q1).with_conditions('a'),
+            )
+        )
+    )
+
+    cirq.testing.assert_has_diagram(
+        circuit,
+        """
+      Circuit_0x0000000000000000:
+      [ 0: ───M───────          ]
+0: ───[       ║                 ]───
+      [ 1: ───╫───X───          ]
+      [       ║   ║             ]
+      [ a: ═══@═══^═══          ]
+      │
+1: ───#2────────────────────────────
+""",
+        use_unicode_characters=True,
+    )
+
+
+def test_diagram_subcircuit_layered():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(
+        cirq.measure(q0, key='a'),
+        cirq.CircuitOperation(
+            cirq.FrozenCircuit(
+                cirq.measure(q0, key='a'),
+                cirq.X(q1).with_conditions('a'),
+            ),
+        ),
+        cirq.X(q1).with_conditions('a'),
+    )
+
+    cirq.testing.assert_has_diagram(
+        circuit,
+        """
+          Circuit_0x0000000000000000:
+          [ 0: ───M───────          ]
+0: ───M───[       ║                 ]───────
+      ║   [ 1: ───╫───X───          ]
+      ║   [       ║   ║             ]
+      ║   [ a: ═══@═══^═══          ]
+      ║   ║
+1: ───╫───#2────────────────────────────X───
+      ║   ║                             ║
+a: ═══@═══╩═════════════════════════════^═══
+""",
+        use_unicode_characters=True,
+    )
+
+
+def test_qasm():
+    q0, q1 = cirq.LineQubit.range(2)
+    circuit = cirq.Circuit(cirq.measure(q0, key='a'), cirq.X(q1).with_conditions('a'))
+    qasm = cirq.qasm(circuit)
+    assert (
+        qasm
+        == """// Generated from Cirq v0.14.0.dev
+
+OPENQASM 2.0;
+include "qelib1.inc";
+
+
+// Qubits: [0, 1]
+qreg q[2];
+creg m_a[1];
+
+
+measure q[0] -> m_a[0];
+if (m_a!=0) x q[1];
+"""
+    )
+
+
```

Comment:
There are some negative tests I'd like to see added here:
- Condition on a measurement key that has not been measured yet
- Inside a subcircuit, condition on a measurement key outside that subcircuit
- Tests around conditional measurement (I assume there's potential for misbehavior if you do `cirq.measure(q0, key='m').with_conditions('a')`, especially if `'a'` is zero)

If you can think of anything of tests of this type (i.e. testing unsupported or invalid behaviors), please add them as well.

Reply:
Sure. I can tell you what the behaviors will be:

> Condition on a measurement key that has not been measured yet
Runtime error, key not found. I'll add a better error message.

> Inside a subcircuit, condition on a measurement key outside that subcircuit
Even though it's not officially supported, this will magically hook up iff there are no repetition ids (runtime error otherwise). I propose just letting this behavior be "undefined" for now. All that is fully fleshed out and tested in the scope PR that's next in line.

> Tests around conditional measurement
This would behave as expected (like the gate was there or it was not), but I'll add a test. Though that does bring up whether to even allow this. My gut feeling is no, at least initially. Thoughts?

Reply:
I also think we should prohibit any of these as ControlledOperation subops.

Reply:
Agreed on prohibiting `ConditionalOperation`s as `ControlledOperation` subops - if possible, they should be prohibited by the same rules that prevents controlled measurement.

I vote for allowing conditional measurement, as it's necessary for some useful constructions:
```python
circuit_a = Circuit(
    cirq.measure(q0, key='k'),
    # if k, run subcircuit with self-contained measurements
    CircuitOperation(FrozenCircuit(
        cirq.measure(q1, key='m'),
        cirq.X(q1).with_conditions('m')
    )).with_conditions('k')
)

circuit_b = Circuit(
    cirq.measure(q0, key='k'),
    # if k, measure q1; else measure q2
    cirq.measure(q1, key='m').with_conditions('k')
    cirq.measure(q2, key='m').with_conditions('k', invert_mask=[True])
)
```

Reply:
Yeah, the "at least not initially" was intentional. The thing that concerns me is
```python
Circuit(
    cirq.measure(q0, key='k'),
    cirq.measure(q1, key='m').with_conditions('k'),
    cirq.X(q1).with_conditions('m')
)
```
, which would be a valid circuit (under the existing logic) but throw a runtime error.

I'd want to make sure we have a design that prevents such circuits from being constructed first. In you first example, maybe we say `cirq.measure(q1, key='m', local_key=True)` or something, and for the second one we make a new `IfElseOperation(op_if, op_else, condition)` that ensures both ops have the same exposed measurement keys.

Or maybe I am overthinking it and we just let users figure it out themselves.

Reply:
+1 to not allowing this in this PR (incl. tests to verify this), and revisiting later.

This essentially boils down to "reference to possibly-uninitialized variable". Coming from C++ and Python where this is permitted (albeit with a warning in C++), I'm in favor of letting users handle this; however, given that there are languages like Java which do not allow this, I can see either choice being valid.

Reply:
FWIW we don't currently have a check against controlled measurements. I will add one. There are probably other things that shouldn't be controlled (reset gate?) but I'll leave that for later.

Reply:
Had to undo the check against controlled measurements. That broke consistent protocols test, and was outside the scope of this PR anyway. We can revisit that later. Added the remaining tests and required functionality to conditional_op though.